ruby/sdk-phase-1.md

```diff
@@ -0,0 +1,439 @@
+
+  # Can customize executor, e.g.
+  # activity_executor :fiber
+
+  # Can make an activity dynamic, e.g.
+  # activity_dynamic true
+
+  def initialize(db_client)
+    @db_client = db_client
+  end
+
+  def execute(arg)
+    loop do
+      db_result = @db_client.do_thing(arg)
+      if db_result
+        return db_result
+      end
+      sleep 1
+      Temporalio::Activity::Context.heartbeat
+    end
+  end
+end
+```
+
+* Activities are defined as classes.
+  * 💭 Why classes instead of methods/functions like every other SDK? There are tradeoffs to both approaches. As
+    classes, they are more easily referenced, a bit more easily typed in RBS, and metadata can be applied more easily at
+    the class level. This comes at a cost of not being able to easily share state across classes, but users can provide
+    something in the constructor multiple activities use. Open to considering activity methods instead.
+  * ❓ Would we prefer activities as methods knowing they may not be able to be typed/referenced well? We can't really
+    take the workflow signal-ish approach of those pseudo-decorators making class methods unless we say activities must
+    be instance methods (since we would allow class methods as activities).
+* Activity classes are instantiated at registration time, not for each invocation.
+  * 💭 Why? Existing Temporal Ruby SDK does instantiate for each activity task, but this discourages use of shared
+    objects or state to be referenced. This does have the tradeoff our other SDKs have of requiring the function to be
+    thread safe.
+* `Temporalio::Activity::Context` is a thread/fiber-local context available to activities.
+  * It has instance methods for info, heartbeat, etc.
+  * It has a class method for `current` that returns from thread/fiber-local and class method shortcuts for all instance
+    methods.
+    * 💭 Why class methods equivalents? `Temporalio::Activity::Context.heartbeat` is easier to read than
+      `Temporalio::Activity::Context.current.heartbeat`. In .NET we made them get current instance first because that's
+      normal there, and in Python we only offered package-level methods because that's normal there.
+    * 💭 Why allow access to an instance? Some users want to pass the activity context to another thread.
+  * 💭 Why `Temporalio::Activity::Context` instead of just `Temporalio::Activity` or putting helpers on base class or
+    something? It is a common requirement in our SDKs to be able to access the activity context from the environment.
+    Also, it is a common requirement to be able to do that from utility methods outside of the explicit activity call.
+* Activity can choose its executor, default is `:default`.
+  * See worker section for more information on executors.
+* Cancellation is reported to the executor.
+  * For thread pool executor (the default), this raises a cancellation exception via `Thread.raise`.
+  * For fiber executor, this is done via `Fiber.raise`.
```

Comment:
Is there a way to abstract the low level implementation (fiber vs thread pool) from how the cancellation is reported to the activity code. Ideally, we would like to change the implementation without modifying application code...

Reply:
+1 - would be nice to just have `Temporalio::Activity::Context.raise`

Reply:
> Is there a way to abstract the low level implementation (fiber vs thread pool) from how the cancellation is reported to the activity code

It is abstracted from a user POV. This discusses what _our_ SDK code will do when these happen. From a user POV it's just an exception that occurs.

Reply:
Ah, cool that makes sense, that wasn't immediately clear to me.